### PR TITLE
avoid op hashing in param-shift-cv and metric-tensor

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
Op hashing raises warnings-turned-errors in tests, and we should avoid doing it.

**Description of the Change:**
- Update parameter_shift_cv to not use `list_of_ops.index()`. This never caused warnings because the index is always 0. I found this funny - I guess python first checks `if op1 is op2`, and if that fails, it tries `if hash(op1) == hash(op2)` when indexing a list
- Update `metric_tensor` to use a `WrappedObj`. `_get_gen_op` uses `functools.lru_cache`, and that uses the parameter hashes (which includes `Operator.__hash__`) to check for a cache hit.

**Benefits:**
We depend less on operator hashing!

**Possible Drawbacks:**
We'll probably want to undo these changes in the future.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -318,6 +318,7 @@ array([False, False])
   warnings to reflect upcoming changes to operator and measurement process equality and hashing.
   [(#4144)](https://github.com/PennyLaneAI/pennylane/pull/4144)
   [(#4454)](https://github.com/PennyLaneAI/pennylane/pull/4454)
+  [(#)](https://github.com/PennyLaneAI/pennylane/pull/)
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:321]:
```suggestion
  [(#4489)](https://github.com/PennyLaneAI/pennylane/pull/4489)
```

**Review by mudit2812 — APPROVED** [review]
Looks good.
Re: measurement process hashing, there are a few places where the new hash is what we want, so we need to be careful about adding `WrappedObj` or other workarounds to "wrap" the hash.

File: pennylane/gradients/parameter_shift_cv.py [hunk 2]
```diff
@@ -396,8 +396,11 @@ def second_order_param_shift(tape, dev_wires, argnum=None, shifts=None, gradient
 
         for obs in observable_descendents:
             # get the index of the descendent observable
-            idx = tape.observables.index(obs)
-            transformed_obs_idx.append(idx)
+            # pylint:disable=undefined-loop-variable
+            for obs_idx, tape_obs in enumerate(tape.observables):
+                if qml.equal(obs, tape_obs):
```
Comment (timmysilv) [inline pennylane/gradients/parameter_shift_cv.py:401]:
```suggestion
                if obs is tape_obs:
```

File: pennylane/gradients/parameter_shift_cv.py [hunk 3]
```diff
@@ -396,8 +396,11 @@ def second_order_param_shift(tape, dev_wires, argnum=None, shifts=None, gradient
 
         for obs in observable_descendents:
             # get the index of the descendent observable
-            idx = tape.observables.index(obs)
-            transformed_obs_idx.append(idx)
+            # pylint:disable=undefined-loop-variable
+            for obs_idx, tape_obs in enumerate(tape.observables):
+                if qml.equal(obs, tape_obs):
+                    break
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/gradients/parameter_shift_cv.py]: probably `is`. updated

**Review by albi3ro — APPROVED** [review]

**Pull request merged**